src/components/CompanyAndJobTitle/CompanyAndJobTitleWrapper.js

```diff
@@ -1,17 +1,85 @@
 import React, { useMemo } from 'react';
+import { generatePath, useRouteMatch } from 'react-router';
+import { useSelector } from 'react-redux';
 import PropTypes from 'prop-types';
 import { toPairs, compose, map } from 'ramda';
 
 import Heading from 'common/base/Heading';
 import FanPageBlock from 'common/FanPageBlock';
 import BreadCrumb from 'common/BreadCrumb';
+import Redirect from 'common/routing/Redirect';
 
-import { tabTypeTranslation, generateTabURL } from 'constants/companyJobTitle';
+import {
+  tabTypeTranslation,
+  generateTabURL,
+  pageType as PAGE_TYPE,
+  tabType as TAB_TYPE,
+} from 'constants/companyJobTitle';
 import { generateBreadCrumbData } from './utils';
 
+import {
+  name as nameSelectorFromBox,
+  company as companyBoxSelectorByPageName,
+  jobTitle as jobTitleBoxSelectorByPageName,
+  companyOverviewBoxSelectorByName,
+  jobTitleOverviewBoxSelectorByName,
+} from 'selectors/companyAndJobTitle';
+
 import TabLinkGroup from 'common/TabLinkGroup';
 import styles from './CompanyAndJobTitleWrapper.module.css';
 
+const useBoxSelector = ({ pageType, pageName, tabType }) => {
+  switch (pageType) {
+    case PAGE_TYPE.COMPANY:
+      if (tabType === TAB_TYPE.OVERVIEW) {
+        return companyOverviewBoxSelectorByName(pageName);
+      }
+      return companyBoxSelectorByPageName(pageName);
+
+    case PAGE_TYPE.JOB_TITLE:
+      if (tabType === TAB_TYPE.OVERVIEW) {
+        return jobTitleOverviewBoxSelectorByName(pageName);
+      }
+      return jobTitleBoxSelectorByPageName(pageName);
+
+    default:
+      return null;
+  }
+};
+
+const useCanonicalPath = ({ pageType, name }) => {
+  const { path } = useRouteMatch();
+
+  if (!name) return null;
+
+  switch (pageType) {
+    case PAGE_TYPE.COMPANY:
+      return generatePath(path, { companyName: name });
+
+    case PAGE_TYPE.JOB_TITLE:
+      return generatePath(path, { jobTitle: name });
+
+    default:
+      return null;
+  }
+};
+
+const useRedirectPath = ({ pageType, pageName, tabType }) => {
+  const boxSelector = useBoxSelector({ pageType, pageName, tabType });
+  const nameSelector = compose(
+    nameSelectorFromBox,
+    boxSelector,
+  );
```

Comment:
A boxSelector has property {status, data}, there is cases

1. status is not fetched --> should not redirect now, wait status to be fetched
2. data is null --> should not consider redirect
3. here